### PR TITLE
TSX: Fix incorrectly removing parentheses around jsx elements being accessed in tsx file

### LINKED ISSUE #5683: TSX parser incorrectly removes parentheses around inline React elements being accessed
Reproduce with this code:

```ts
import React from "react"

const Example  = () => {
    const foo = (<a/>).toString()
    return <a/>
}
```

This removes the parentheses, creating the invalid syntax `foo = <a/>.toString()`.

### PR DESCRIPTION
**Pull request by sosukesuzuki**
Fix #5683

- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] (If the change is user-facing) I’ve added my changes to the `CHANGELOG.unreleased.md` file following the template.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: src/language-js/needs-parens.js [hunk 1]
```diff
@@ -6,8 +6,9 @@ const util = require("../common/util");
 const comments = require("./comments");
 const {
   getLeftSidePathName,
+  hasFlowShorthandAnnotationComment,
   hasNakedLeftSide,
-  hasFlowShorthandAnnotationComment
+  isTsx
```
Comment (fisker) [inline src/language-js/needs-parens.js]:
```suggestion
  isTsxFile
```

**Comment by thorn0** [comment]
The issue contains a discussion, and [this comment](https://github.com/prettier/prettier/issues/5683#issuecomment-450383220) seems to sum it up. Let's always add parens irrespective of file extension if the parent node is unusual for JSX.

**Comment by sosukesuzuki** [comment]
@thorn0 Thanks. Modified to pass our snapshot tests without depending on the file extension.

**Comment by thorn0** [comment]
Let's do the same for JSXFragment, see [playground](https://deploy-preview-6640--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuEBLAtgBwgJxgAgCU4BDMAgMxwnXwB0QdTyG6o3IoBnAgUQA8SWADZx8+ALz4AFAEpJAPnzA24zj3wUIESTIA8CvQHoFsgHQwIAZRg5UUAOZzV4-ExgBXHFHx6SJtgBfEAAaEAhMGFRoLmRQEhxqAHcABQSEWJQSADcIVAATUJAAIxwyAGs4GCtMMnsHZFsPODCACxh0YQB1VtR4LlqwOCsMvtRsvoBPZHAuWLD7Ljg8FLKHdBJkChJhJbCAKy5+ACEysErqoTgAGXs4LZ29kEP+K3rRAEUPCHgH3ZaQLUcEscDMYJNMHAuGA7JEipg7LAugUYK1kAAOAAMYQRECWXTKmBmCKhy2y9zCAEdvvBVhFMiASFwALRQOBwfIcopMamoJirEjrTZIbb-MJLdCoRo4Zri95wL4-e4ix4AmAkYrI-Ko5AAJjCthIqGE9QAwjQNjMoNAKSAPEsACoazKip7ZZoASSgnNgVhhqEiAEFvTZJqI-ktAoEgA).

File: CHANGELOG.unreleased.md [hunk 2]
```diff
@@ -859,6 +859,22 @@ function doSmth() {
 }
 ```
 
+#### TypeScript: Fix incorrectly removes parentheses around inline jsx elements being accessed in tsx file ([#6640] by [@sosukesuzuki])
```
Comment (thorn0) [inline CHANGELOG.unreleased.md]: Consider changing the wording to "TypeScript: sometimes removing parentheses around JSX made the code unparseable".

Comment (sosukesuzuki) [reply CHANGELOG.unreleased.md]: Thank you, I'm not good at English... It will be helpful if you continue to review.

Comment (thorn0) [reply CHANGELOG.unreleased.md]: Your English is fine, but "fix ... removes parentheses" sounds like your fix removes them, not fixes their removal.

**Comment by sosukesuzuki** [comment]
@evilebottnawi I rebased:+1:

**Comment by sosukesuzuki** [comment]
@evilebottnawi Can I merge this pr?

**Comment by evilebottnawi** [comment]
@sosukesuzuki i think yes

**Comment by slinkardbrandon** [comment]
Seeing a little weirdness with this in my codebase. Basically, it's adding additional nesting of unnecessary parentheses, is this intended?

```tsx
 const wrapper = new ShallowWrapper(
   (<Image type="image" id="1" size="medium" url="http://placecorgi.com/260/260" />)
 ).shallow();
```

```tsx
const wrapper = new ShallowWrapper(
  (
    <Image
      type="image"
      id="1"
      onClick={someFunc|
      url="http://placecorgi.com/260/260"
    />
  )
).shallow();
```

Basically, it seems that any time the JSX/TSX (tsx in my case) wraps, this adds duplicated parentheses.

Is this intentional? If so, is it configurable?

**Comment by thorn0** [comment]
@slinkardbrandon It's a bug. Please open an issue or (better) a PR with a fix for it. Please don't write in closed issues and PRs.

**Comment by lipis** [comment]
@thorn0 Let's not ask people to submit PRs instead of issues.. I guess if they can they will submit it, issues is more than enough.

**Comment by thorn0** [comment]
Ooops, I forgot that this has been fixed: https://github.com/prettier/prettier/pull/6980

@lipis Sometimes bugs can look more complicated than they are. It's in such cases that I usually try encouraging people to submit PRs. And it actually worked out with this specific bug. :) See https://github.com/prettier/prettier/issues/6969#issuecomment-554563757

**Comment by lipis** [comment]
@thorn0 Fair enough :)

**Comment by slinkardbrandon** [comment]
Sorry about that. I didn’t want to open an issue because I didn’t know if it was intentional behavior. Thanks for the update though!

**Pull request merged**